src/extension/ip_extraction.c

```diff
@@ -0,0 +1,596 @@
+    DD_INI_ENV("ipheader", "", PHP_INI_SYSTEM, _on_update_ipheader),
+    {0}
+};
+// clang-format on
+
+static ZEND_INI_MH(_on_update_ipheader)
+{
+    ZEND_INI_MH_UNUSED();
+    if (_ipheader) {
+        zend_string_release(_ipheader);
+    }
+
+    if (!new_value || !ZSTR_VAL(new_value)[0]) {
+        _ipheader = NULL;
+        return SUCCESS;
+    }
+
+    size_t key_len = sizeof("HTTP_") - 1 + ZSTR_LEN(new_value);
+
+    zend_string *normalized_value = zend_string_alloc(key_len, 1);
+    char *out = ZSTR_VAL(normalized_value);
+    memcpy(out, ZEND_STRL("HTTP_"));
+    out += sizeof("HTTP_") - 1;
+    const char *end = ZSTR_VAL(new_value) + ZSTR_LEN(new_value);
+    for (const char *p = ZSTR_VAL(new_value); p != end; p++) {
+        char c = *p;
+        if (c >= 'a' && c <= 'z') {
+            c = (char)(c - 'a' + 'A');
+        } else if (c == '-') {
+            c = '_';
+        }
+        *out++ = (char)c;
+    }
+    *out = '\0';
+
+    _ipheader = normalized_value;
+
+    return SUCCESS;
+}
+
+static void _register_testing_objects(void);
+
+void dd_ip_extraction_startup()
+{
+    dd_phpobj_reg_ini_envs(ini_settings);
+    _x_forwarded_for_key =
+        zend_string_init_interned(ZEND_STRL("HTTP_X_FORWARDED_FOR"), 1);
+    _x_real_ip_key = zend_string_init_interned(ZEND_STRL("HTTP_X_REAL_IP"), 1);
+    _client_ip_key = zend_string_init_interned(ZEND_STRL("HTTP_CLIENT_IP"), 1);
+    _x_forwarded_key =
+        zend_string_init_interned(ZEND_STRL("HTTP_X_FORWARDED"), 1);
+    _x_cluster_client_ip_key =
+        zend_string_init_interned(ZEND_STRL("HTTP_X_CLUSTER_CLIENT_IP"), 1);
+    _forwarded_for_key =
+        zend_string_init_interned(ZEND_STRL("HTTP_FORWARDED_FOR"), 1);
+    _forwarded_key = zend_string_init_interned(ZEND_STRL("HTTP_FORWARDED"), 1);
+    _via_key = zend_string_init_interned(ZEND_STRL("HTTP_VIA"), 1);
+    _true_client_ip_key =
+        zend_string_init_interned(ZEND_STRL("HTTP_TRUE_CLIENT_IP"), 1);
+    _remote_addr_key = zend_string_init_interned(ZEND_STRL("REMOTE_ADDR"), 1);
+
+    _register_testing_objects();
+}
+
+static zend_string *nullable _fetch_arr_str(
+    const zval *nonnull server, zend_string *nonnull key);
+static bool _is_private(const ipaddr *nonnull addr);
+static zend_string *nullable _ipaddr_to_zstr(const ipaddr *ipaddr);
+static zend_string *_try_extract(
+    const zval *nonnull server, zend_string *key, extract_func_t extract_func);
+static bool _parse_x_forwarded_for(
+    zend_string *nonnull value, ipaddr *nonnull out);
+static bool _parse_plain(zend_string *nonnull zvalue, ipaddr *nonnull out);
+static bool _parse_forwarded(zend_string *nonnull zvalue, ipaddr *nonnull out);
+static bool _parse_via(zend_string *nonnull zvalue, ipaddr *nonnull out);
+
+zend_string *nullable dd_ip_extraction_find(zval *nonnull server)
+{
+    zend_string *res;
+
+    if (_ipheader) {
+        zend_string *value = _fetch_arr_str(server, _ipheader);
+        if (!value) {
+            return NULL;
+        }
```

Comment:
I've rechecked the RFC and indeed that may be the best reading. I don't think that that is the best option though, because it leaves the users more exposed to header spoofing, even when they set the header ini value. Given the ambiguity, I'd keep this option. I've also added a comment section on the RFC, hopefully other people can weigh in.